Bluetooth: Mesh: Miscellaneous corrections to minor problems

LingaoM opened the pull request:
Fixed the problem in the Provisioner provisioning
formatted printing and added key update function for CDB.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

trond-snekvik commented:
cc @haoyanbiao

haoyanbiao reviewed (approved):
LGTM

jhedberg commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -183,6 +183,36 @@ int bt_mesh_cfg_gatt_proxy_set(uint16_t net_idx, uint16_t addr, uint8_t val,
+ */
+int bt_mesh_cfg_node_identity_set(uint16_t net_idx, uint16_t addr,
+				  uint16_t key_net_idx, uint8_t val,
+				  uint8_t *status, uint8_t *identity);
```
I don't think we have the habit of returning the "current value" parameter for any other "set" client APIs, so I don't think we should do it here either, i.e. you should remove the last parameter.

jhedberg commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -801,6 +929,47 @@ int bt_mesh_cfg_mod_sub_get_vnd(uint16_t net_idx, uint16_t addr, uint16_t elem_a
+ */
+int bt_mesh_cfg_krp_set(uint16_t net_idx, uint16_t addr,
+			uint16_t key_net_idx, uint8_t val,
+			uint8_t *status, uint8_t *phase);
```
Same thing here regarding not having the "current value" parameter for `set()` functions

jhedberg commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -801,6 +929,47 @@ int bt_mesh_cfg_mod_sub_get_vnd(uint16_t net_idx, uint16_t addr, uint16_t elem_a
+ *  @return 0 on success, or (negative) error code on failure.
+ */
+int bt_mesh_cfg_krp_get(uint16_t net_idx, uint16_t addr,
+			uint16_t key_net_idx, uint8_t *status, uint8_t *phase);
```
Is it a phase or a state? You use both in the function documentation.

trond-snekvik commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -2700,26 +3257,36 @@ SHELL_STATIC_SUBCMD_SET_CREATE(mesh_cmds,
 		      cmd_relay, 1, 3),
+	SHELL_CMD_ARG(mod-pub-va, NULL, "<Label UUID> <mod id> [cid] [<PubAddr> "
+		      "<AppKeyIndex> <cred: off, on> <ttl> <period> <count> <interval>]",
+		      cmd_mod_pub_va, 10, 1),
```
```suggestion
		      cmd_mod_pub_va, 3, 8),
```

trond-snekvik commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -2659,16 +3208,13 @@ SHELL_STATIC_SUBCMD_SET_CREATE(mesh_cmds,
 
 	/* Commands which access internal APIs, for testing only */
-	SHELL_CMD_ARG(net-send, NULL, "<hex string>", cmd_net_send, 2, 0),
+	SHELL_CMD_ARG(net-send, NULL, "<ttl> <data>", cmd_net_send, 3, 0),
```
There's already a `ttl` command that configures the default TTL value, use that instead of changing this command?

trond-snekvik commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -2000,13 +2540,22 @@ int cmd_timeout(const struct shell *shell, size_t argc, char *argv[])
 		}
 
 		bt_mesh_cfg_cli_timeout_set(timeout_ms);
+		bt_mesh_health_cli_timeout_set(timeout_ms);
```
Please change the description for this command in doc/reference/bluetooth/mesh/shell.rst to show that this now also changes the health client timeout

trond-snekvik commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -1932,6 +2350,128 @@ static int cmd_provision_adv(const struct shell *shell, size_t argc,
+		"84DBC966204796ECCF0D6CF5E16500CC"
+		"0201D048BCBBD899EEEFC424164E33C2"
+		"01C2B010CA6B4D43A8A155CAD8ECB279"
+	};
```
I don't know how useful this will be. The public key should be different for every device, but this locks our shell application image to a single key. This command should at least accept a parameter to set this key

trond-snekvik commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -1932,6 +2350,128 @@ static int cmd_provision_adv(const struct shell *shell, size_t argc,
+		size = str2u8(argv[2]);
+
+		err = bt_mesh_auth_method_set_output(act, size);
+	}
```
Remove the `else` in these `return ... else` statements, so we can keep the `else`-block at the innermost indentation

trond-snekvik commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -1932,6 +2350,128 @@ static int cmd_provision_adv(const struct shell *shell, size_t argc,
+static int cmd_prov_close_link(const struct shell *shell, size_t argc,
+			     char *argv[])
+{
+	pb_adv.link_close(PROV_BEARER_LINK_STATUS_FAIL);
```
Shouldn't bypass the provisioning API like this. Add a function to prov.h instead, so we don't get a bug here when we add PB-GATT

trond-snekvik commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -1721,15 +2018,94 @@ static int cmd_mod_pub(const struct shell *shell, size_t argc, char *argv[])
+	mod_id = strtoul(argv[2], NULL, 0);
+
+	argc -= 3;
+	argv += 3;
```
This seems like a potential future bug, why not keep do `argv[3]` below?

LingaoM replied:
?

trond-snekvik commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -1721,15 +2018,94 @@ static int cmd_mod_pub(const struct shell *shell, size_t argc, char *argv[])
+		return -EINVAL;
+	}
+	interval = strtoul(argv[6], NULL, 0);
+	if (interval > (31 * 50) || (interval % 50)) {
```
Use rounding for the interval instead? This looks like it's going to be hard to get right without looking at the spec or source code

trond-snekvik commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -896,6 +917,31 @@ static int cmd_gatt_proxy(const struct shell *shell, size_t argc, char *argv[])
+		uint16_t lpn_address = strtoul(argv[1], NULL, 0);
+		err = bt_mesh_cfg_lpn_timeout_get(net.net_idx, net.dst,
+						  lpn_address, &timeout);
+	}
```
Replace `return else ` here as well

trond-snekvik commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -1868,6 +2123,197 @@ int bt_mesh_cfg_mod_pub_set_vnd(uint16_t net_idx, uint16_t addr, uint16_t elem_a
+	}
+
+	return cli_wait();
+}
```
Return error on invalid phases? (`phase <= 1 || phase > 3`)

LingaoM replied:
Can let the server model do this check?

trond-snekvik replied:
It will, but the API shouldn't allow illegal values to go on air

trond-snekvik reviewed (changes requested):
All the new commands in the shell needs their own entries in doc/reference/bluetooth/mesh/shell.rst. Please add them in this PR, so we can have a round of reviews on those as well.

LingaoM commented on include/bluetooth/mesh/health_cli.h:
```diff
@@ -62,14 +62,6 @@ struct bt_mesh_health_cli {
- *
- *  @return 0 on success, or (negative) error code on failure.
- */
-int bt_mesh_health_cli_set(struct bt_mesh_model *model);
```
It is not used anywhere. Is it necessary to exist?

trond-snekvik replied:
I believe so, it makes it possible to have multiple health clients

LingaoM commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -938,6 +1124,79 @@ int bt_mesh_cfg_net_transmit_get(uint16_t net_idx, uint16_t addr,
+		return err;
+	}
+
+	return cli_wait();
```
This is a get function. If you add reset, it will be meaningless?

trond-snekvik replied:
Sorry, you're right!

LingaoM commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -938,6 +1124,79 @@ int bt_mesh_cfg_net_transmit_get(uint16_t net_idx, uint16_t addr,
+		cli_reset();
+		return err;
+	}
+
```
It is necessary to add reset here

LingaoM committed:
Bluetooth: Mesh: Fix provisioner provisioning

Correct format errors, such as oob action 2-octers
should use 0x%04x, and action in prov capalilities pdu
big-ending. When every call `bt_mesh_auth_method_set<*>`
should also clear auth value, otherwise will case confirm
failed.

Provisioner role expect peer public key should be sent
immediately, instead of requiring ACK. After all, ACK may
be lost, and the other device’s public key will be sent
over, and provisioning procedure will be failed..

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Add CDB handle key refresh phase

Fixed some undesired or incorrect macros, which
can cause build error.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Fix cfg_cli Format and CID use problem

Fix some format warning in cfg_cli.c and use CID_VNAL as
param when use vendor models.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Fix print format and argument resolve

Add line breaks to make the results more intuitive
Some parameter analysis problems are corrected.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
I have submitted all the units separately. I suggest merging the current units first. After all, they will solve some problems in the current code, and then I will complete them in another pr.

trond-snekvik commented on subsys/bluetooth/mesh/provisioner.c:
```diff
@@ -667,6 +669,8 @@ static void prov_set_method(uint8_t method, uint8_t action, uint8_t size)
 	bt_mesh_prov_link.oob_action = action;
 	bt_mesh_prov_link.oob_size = size;
+
+	(void)memset(bt_mesh_prov_link.auth, 0, sizeof(bt_mesh_prov_link.auth));
```
If you really want to move this out of set_input, I believe it needs to go in reset_state, or something else that's always called between provisioning sessions. The application doesn't have to call this for every run, which means that the input oob value may have dirty data at the end if you provision multiple devices without setting the oob method every time.

trond-snekvik reviewed (changes requested):
I have marked the comments I made on the removed content as resolved, but I hope you will take them into consideration in your separate PR.

There's just one comment I don't think has been addressed yet, see above. I don't see the usecase for setting KR phase = 1 in CDB, I think it should be implicit when you start adding keys, as the spec does. Is there any reason to set the KR phase separately?

I have one additional comment.

jhedberg commented on subsys/bluetooth/mesh/provisioner.c:
```diff
@@ -698,10 +700,7 @@ int bt_mesh_auth_method_set_static(const uint8_t *static_val, uint8_t size)
 	prov_set_method(AUTH_METHOD_STATIC, 0, 0);
 
 	memcpy(bt_mesh_prov_link.auth + 16 - size, static_val, size);
-	if (size < 16) {
-		(void)memset(bt_mesh_prov_link.auth, 0,
-			     sizeof(bt_mesh_prov_link.auth) - size);
-	}
```
Why is this part being removed? I don't see an explanation of it in the commit message (or then I didn't understand that part in the commit message).

LingaoM replied:
When every call `bt_mesh_auth_method_set<*>`
should also clear auth value, otherwise will case confirm
failed.

jhedberg replied:
@LingaoM sorry, I'm still not understanding. First of all, I'm not seeing any clearing in the other set() methods like you say. Secondly, all this memset was doing was to ensure that no garbage is left in `bt_mesh_prov_link.auth` if the new size is smaller than the previous one.

carlescufi replied:
This function, or indeed others in the `bt_mesh_auth_method_set_*` do **not** clear the `auth` array.

LingaoM commented:
Close this, please move to https://github.com/zephyrproject-rtos/zephyr/pull/32567
